Tests/ApolloTests/Network/URLSessionClientTests.swift

```diff
@@ -337,8 +377,221 @@ class URLSessionClientTests: XCTestCase {
+    // and that may handle the callbacks differently.
+    //
+    // 1. When multipart chunks are received, to be processed immediately - from urlSession(_:dataTask:didReceive:)
+    // 2. When the operation completes, with any remaining task data - from urlSession(_:task:didCompleteWithError:)
+    expectation.expectedFulfillmentCount = 2
+
+    self.client.sendRequest(request) { result in
+      switch result {
+      case .failure(let error):
+        XCTFail("Unexpected error: \(error)")
+
+      case .success(let (data, httpResponse)):
+        XCTAssertTrue(httpResponse.isSuccessful)
+        XCTAssertTrue(httpResponse.isMultipart)
+
+        switch String(data: data, encoding: .utf8) {
+        case completeChunk, partialChunk:
+          expectation.fulfill()
+
+        default:
+          XCTFail("Unexpected data received: \(data)")
+        }
+      }
+    }
+
+    self.wait(for: [expectation], timeout: 1)
+  }
+
+  func test__multipart__givenChunkContainingBoundaryString_shouldNotSplitChunk() throws {
+    let url = URL(string: "http://www.test.com/multipart-containing-boundary-string")!
+    let boundary = "-"
+    let multipartString = "--\(boundary)\r\nContent-Type: application/json\r\n\r\n{\"data\": {\"field1\": \"value1--\(boundary)\"}}\r\n--\(boundary)--"
+
+    let request = self.request(
+      for: url,
+      responseData: multipartString.data(using: .utf8),
+      statusCode: 200,
+      headerFields: ["Content-Type": "multipart/mixed; boundary=\(boundary)"]
+    )
+
+    let expectation = self.expectation(description: "Multipart chunk received")
+
+    // Results are sent twice for multipart responses with both received here because this test infrastructure uses
+    // URLSessionClient direclty whereas in a request chain it is wrapped by an interceptor (NetworkFetchInterceptor)
+    // and that may handle the callbacks differently.
+    //
+    // 1. When multipart chunks are received, to be processed immediately - from urlSession(_:dataTask:didReceive:)
+    // 2. When the operation completes, with any remaining task data - from urlSession(_:task:didCompleteWithError:)
+    expectation.expectedFulfillmentCount = 2
+
+    self.client.sendRequest(request) { result in
+      switch result {
+      case .failure(let error):
+        XCTFail("Unexpected error: \(error)")
+
+      case .success(let (data, httpResponse)):
+        XCTAssertTrue(httpResponse.isSuccessful)
+        XCTAssertTrue(httpResponse.isMultipart)
+
+        switch String(data: data, encoding: .utf8) {
+        case multipartString, "": // "" is the second result and is expected to be empty
+          expectation.fulfill()
+
+        default:
+          XCTFail("Unexpected data received: \(data)")
+        }
+      }
+    }
+
+    self.wait(for: [expectation], timeout: 1)
+  }
+
+  func test__multipart__givenChunkContainingBoundaryStringWithoutClosingBoundary_shouldNotSplitChunk() throws {
```

Comment:
This doesn't appear to actually test the critical path.

I was a bit confused about how this managed to work, but when I put a breakpoint on L581, we're actually getting the callback from urlSession dataTask didCompleteWithError and not urlSession dataTask didReceive. Which means we're bypassing the problematic parsing code.

From what I could see earlier, this behavior is actually time sensitive, making it hard to control which parts of URLSessionClient actually get exercised. Which is why I ended up calling its data handler directly.

Reply:
> I was a bit confused about how this managed to work, but when I put a breakpoint on L581, we're actually getting the callback from urlSession dataTask didCompleteWithError and not urlSession dataTask didReceive. Which means we're bypassing the problematic parsing code.

This is an implementation detail; there is no guarantee the different callbacks nor the logic within them will remain the same in the future. The test message data is what's important and there are multiple tests to cover different combinations of the test data.
- `test__parsing__givenSingleChunk_withDashBoundaryInMessageBody_shouldReturnSuccess`
- `test__multipart__givenChunkContainingBoundaryString_shouldNotSplitChunk`
- `test__multipart__givenChunkContainingBoundaryStringWithoutClosingBoundary_shouldNotSplitChunk`

Reply:
> From what I could see earlier, this behavior is actually time sensitive, making it hard to control which parts of URLSessionClient actually get exercised. Which is why I ended up calling its data handler directly.

It's not time sensitive but dependent on the type of response (multipart header) and number of chunks. As data is sent in it's checked for complete chunks so they can be processed immediately; this is the callback that needs to check chunk boundaries. Then once the request completes any left over response data is sent on for further parsing.